### PR TITLE
*: support out-of-order time window for Prometheus

### LINKED ISSUE #5052: Configurable TSDB OutOfOrderTimeWindow
**What is missing?**

While it hasn't been released yet, the next version of Prometheus will support configuring an out of order window for the TSDB.
Configuring this should be available in the Prometheus CR.

**Why do we need it?**

This helps immensely for clouds that silently migrate VMs and cause clock hiccups.
Without this change, a single out of order timestamp indefinitely locks up Prometheus.

**Environment**

* Prometheus Operator version: HEAD

**Anything else we need to know?**:

https://github.com/prometheus/prometheus/commit/c1b669bf9b0b1286ece53c977262089119783105
https://github.com/prometheus/prometheus/issues/11329
https://github.com/prometheus/prometheus/pull/11340

### PR DESCRIPTION
**Pull request by simonpasquier**
## Description

Fixes #5052 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [X] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Added `spec.tsdb.outOfOrderTimeWindow` field to the Prometheus CRD to allow out-of-order samples in TSDB.
```

### CODE REVIEW

File: pkg/prometheus/promcfg.go [hunk 1]
```diff
@@ -626,9 +626,24 @@ func (cg *ConfigGenerator) Generate(
...
+
 	return yaml.Marshal(cfg)
 }
 
+func (cg *ConfigGenerator) appendTSDBConfig(cfg yaml.MapSlice, tsdb v1.TSDBSpec) yaml.MapSlice {
+	if tsdb.OutOfOrderTimeWindow == "" {
+		return cfg
+	}
+
+	return cg.WithMinimumVersion("2.29.0").AppendMapItem(cfg, "tsdb", yaml.MapSlice{
```
Comment (slashpai) [inline pkg/prometheus/promcfg.go:639]: This should be `2.39.0`?

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:639]: :facepalm:

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:639]: thanks!

**Review by PhilipGough — APPROVED** [review]
lgtm

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -626,9 +626,24 @@ func (cg *ConfigGenerator) Generate(
...
 	return yaml.Marshal(cfg)
 }
 
+func (cg *ConfigGenerator) appendTSDBConfig(cfg yaml.MapSlice, tsdb v1.TSDBSpec) yaml.MapSlice {
+	if tsdb.OutOfOrderTimeWindow == "" {
+		return cfg
+	}
+
+	return cg.WithMinimumVersion("2.39.0").AppendMapItem(cfg, "tsdb", yaml.MapSlice{
+		{
```
Comment (slashpai) [inline pkg/prometheus/promcfg.go:640]: Should we log warning to user if lower prometheus version is used and `OutOfOrderTimeWindow` specified?

Comment (simonpasquier) [reply pkg/prometheus/promcfg.go:640]: it's already taken care of by `cg.WithMinimumVersion("2.39.0")` :)

Comment (slashpai) [reply pkg/prometheus/promcfg.go:640]: aah thanks :)

**Pull request merged**